File: islandPerimeter.cpp

// leetcode.com 463. Island Perimeter

class Solution {
public:
    int islandPerimeter(vector<vector<int>>& grid) {
        if(grid.empty() || grid[0].empty()) return 0 ;
        
        int ans = 0 ;
        
        int m = grid.size(), n = grid[0].size() ;
        
        for(int x = 0 ; x < m ; ++x)
        {
            for(int y = 0 ; y < n ; ++y)
            {
                if(grid[x][y] == 1) 
                {
                    ans += 4 ;
                
                    if(x > 0 && grid[x-1][y] == 1)  ans -= 2 ;
                
                    if(y > 0 && grid[x][y-1] == 1)  ans -= 2 ;
                }
            }
        }
        
        return ans ;
    }
};

File: leetcode/java/src/main/java/com/vkeonline/leetcode/p400/IslandPerimeter.java
package com.vkeonline.leetcode.p400;

/**
 * Leetcode [M]: 463. Island Perimeter
 * @author csgear
 */
public class IslandPerimeter {
    public int islandPerimeter(int[][] grid) {
        int perimeter = 0;
        int r = grid.length, c = grid[0].length;

        for (int i = 0; i < r; ++i) {
            for (int j = 0; j < c; ++j) {
                if (grid[i][j] == 0) {
                    continue;
                }
                perimeter += 4;
                if (i > 0) {
                    perimeter -= grid[i - 1][j];
                }
                if (j > 0) {
                    perimeter -= grid[i][j - 1];
                }
                if (i < r - 1) {
                    perimeter -= grid[i + 1][j];
                }
                if (j < c - 1) {
                    perimeter -= grid[i][j + 1];
                }
            }
        }
        return perimeter;

    }
}
